### ISSUE TITLE
Add `epoch_logs` and `training_logs` variables from Model.fit as class attributes

### ISSUE THREAD

**Issue by qmarcou** [keras-team/tf-keras#565]
**System information**.

TensorFlow version (you are using): 2.8.0
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

Use-case: I'm using the keras-tuner library to optimize hyperparameters, but I also want to be able to use Early stopping. In order to get dataset orthogonalization right I'd like to be able to divide my dataset in 4 splits:
- train: on which the actual training is made
- train_val: which is used for Early stopping
- dev_val or validation: used to monitor target metrics to tune hyperparameters
- test: well to test my final best model

I want to make a clear separation between train_val and dev_val so as not to overfit the dev_val set due to early stopping. Thus I would like to perform evaluation on 2 different datasets on each epoch end. Model.fit only handles 1 validation set.

Because keras_tuner uses a ModelCheckpoint callback to save the best models, I can't add the evaluation outside the fit method in some kind of wrapper.

If I could instantiate a custom Callback to evaluate my model on my dev_val set and add the results to the `logs` before ModelCheckpoint.on_epoch_end is called my problem would be solved, but the `logs` dict is passed as read only argument and cannot be modified. 
Right now the logs variables are temporary variables inside the model.fit funciton:
```python
val_logs = {'val_' + name: val for name, val in val_logs.items()}
epoch_logs.update(val_logs)

callbacks.on_epoch_end(epoch, epoch_logs)
training_logs = epoch_logs
```

Solving my problem by subclassing the Model class and overriding the fit method to incorporate the train_val evaluation seems very heavy for such a simple task, that should be performed by Callbacks

**Will this change the current api? How?**
This would enable to use custom Callback to update the logs dict while right now these are read only 

**Who will benefit from this feature?**
I guess anybody wanting to compute some set of metrics only at batch end for instance, or on different validation sets
For instance this would be an elegant way to solve this [SO question](https://stackoverflow.com/questions/54634435/how-to-set-keras-custom-metrics-to-only-be-called-at-the-end-of-epoch)
Possibly many other applications for just extending pieces of predict/evaluate/or fit.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): simply add the two variables as Model attributes (~5 lines of code change)
